crdtree/src/CRDTree.ts

```diff
@@ -106,23 +107,41 @@ export class CRDTree<T = any> implements ICRDTree<T> {
 		return [];
 	}
 
-	public fork(): ID {
-		return undefined;
+	public fork(): BranchID {
+		let newBranch: BranchID = uuid();
+		this.forkWithRef(newBranch);
+		this.checkout(newBranch);
+		return newBranch;
 	}
 
-	public join(ref: ID): void {
+	private forkWithRef(ref: BranchID): void {
+		this.makeChange({
+			kind: ActionKind.FORK,
+			parent: this.state.latest(),
+			parentBranch: this.state.getBranchID(),
+		}, ref);
+	}
+
+	public join(ref: BranchID): void {
+		this.makeChange({
+			kind: ActionKind.JOIN,
+			joinedAt: this.state.latestFrom(ref),
+			joinedBranch: ref,
+		});
+		this.checkout(this.state.getBranchID());
```

Comment:
woah what's this for? does this just force a reload of the objects?

Reply:
Yes. Joining invalidates current history since there are possibly unapplied updates from the joined branch, so it's a force reload. Let me know if there's a better way to do it.